### DISCUSSION TITLE
[Error] Unable to create SOAP Factory - Provider not found

### DISCUSSION THREAD

**Discussion by lferna** [quarkiverse/quarkus-cxf#377] [Q&A] [answered]
Hi all,
We are using SOAPFactory to set SoapElements to our Headers. Something like that:

`SOAPFactory sf = SOAPFactory.newInstance();`
`SOAPElement seqElement = sf.createElement(new QName("uri:org.apache.cxf","dummy"));`
`SOAPElement textElement = sf.createElement(new QName("uri:org.apache.cxf","value"));
textElement.addTextNode("decapitated");`
`seqElement.addChildElement(textElement);`
`SoapHeader dummyHeader = new SoapHeader(new QName("uri:org.apache.cxf","dummy"), seqElement);`

It works in local environment (JVM) but it doesn't work in native mode. This is the error:

`Unable to create SOAP Factory: Unable to create SAAJ meta-factory: Provider com.sun.xml.internal.messaging.saaj.soap.SAAJMetaFactoryImpl not found - null`

I've added these dependencies but without any result:

`com.sun.xml.messaging.saaj.saaj-impl`
`javax.xml.soap.saaj-api`

Any ideas? Thanks.

**Comment by shumonsharif** [comment]
Hi @lferna Would it be possible for you to put together a simple reproducer?  This is likely more a general configuration issue than an issue with the extension, but I can help you get to the bottom of it more easily with a sample project.

**Comment by lferna** [comment]
Hi @shumonsharif 
I've just uploaded a simple reproducer but with different SOAP endpoint (I'm using the soap services of my customer) but the error is before the invocation, when I try to set the Soap Headers, using SOAPFactory.

https://github.com/lferna/sample-client-cxf-quarkus.git

My Quarkus version is 2.6.0.Final

Thanks so much in advance

**Comment by shumonsharif** [comment, marked as answer]
Hi @lferna First off, I would suggest you review this Quarkus native application guide, it should hopefully explain the changes and why they are needed.  There are also other guides on GraalVM that you can lookup.
https://quarkus.io/guides/writing-native-applications-tips

There may possibly be some way to address this in the extension, I'll look into it further.  In the meantime, you can try the steps below.

The `saaj-impl` dependency is already pulled in by CXF.  However, `saaj-impl` has an optional dependency on `mimepull`, which you will need to add to your project for the native build.

```xml
	  <dependency>
		  <groupId>org.jvnet.mimepull</groupId>
		  <artifactId>mimepull</artifactId>
		  <version>1.9.14</version>
	  </dependency>
```

You can add `src/main/resources/META-INF/native-image/reflect-config.json`:

```json
[
  {
    "name" : "com.sun.xml.messaging.saaj.soap.SAAJMetaFactoryImpl",
    "allDeclaredConstructors" : true,
    "allPublicMethods" : true
  },
  {
    "name" : "com.sun.xml.messaging.saaj.soap.ver1_1.SOAPFactory1_1Impl",
    "allDeclaredConstructors" : true,
    "allPublicMethods" : true
  },
  {
    "name" : "com.sun.xml.messaging.saaj.soap.ver1_1.SOAPMessageFactory1_1Impl",
    "allDeclaredConstructors" : true,
    "allPublicMethods" : true
  },
  {
    "name" : "com.sun.xml.messaging.saaj.soap.ver1_2.SOAPFactory1_2Impl",
    "allDeclaredConstructors" : true,
    "allPublicMethods" : true
  },
  {
    "name" : "com.sun.xml.messaging.saaj.soap.ver1_2.SOAPMessageFactory1_2Impl",
    "allDeclaredConstructors" : true,
    "allPublicMethods" : true
  },
  {
    "name":"com.sun.xml.messaging.saaj.soap.impl.ElementImpl",
    "methods":[{"name":"<init>","parameterTypes":["com.sun.xml.messaging.saaj.soap.SOAPDocumentImpl","org.w3c.dom.Element"] }]
  }
]
```

and also `src/main/resources/META-INF/native-image/resource-config.json`:

```json
{
  "resources": {
    "includes": [
      {
        "pattern": "META-INF/services/javax.xml.soap.MessageFactory"
      },
      {
        "pattern": "META-INF/services/javax.xml.soap.SAAJMetaFactory"
      },
      {
        "pattern": "META-INF/services/javax.xml.soap.SOAPConnectionFactory"
      },
      {
        "pattern": "META-INF/services/javax.xml.soap.SOAPFactory"
      }
    ]
  }
}
```

That should be all that's needed.  Let me know in case of any issues.

**Comment by lferna** [comment]
Hi @shumonsharif 

First of all, thank you for your reply and for your attention. Probably, this message should have been posted in other forums, but I want to continue solving my problems here.

Secondly, it works!  🎉🎉

On the other hand, before posting this message, I've tried to add the class `com.sun.xml.messaging.saaj.soap.SAAJMetaFactoryImpl` into the file `reflection-config.json` as the documentation says, but I received another error (not the same, but pretty similar):
`Error while searching for service [javax.xml.soap.SOAPFactory] ` I suppose this service is what I have to add in the `resource-config.json` file but I didn't know that. Probably, it's part of my ignorance of how CXF works and which classes should be added in those files.

Once we put these microservices in production environment, I'll let you know about the success story.

Thanks!
